File: src/main/java/org/sonar/plugins/xquery/checks/XPathSubExpressionsInPredicateCheck.java
/*
 * © 2014 by Intellectual Reserve, Inc. All rights reserved.
 */

package org.sonar.plugins.xquery.checks;

import org.sonar.api.rule.RuleKey;
import org.sonar.check.Priority;
import org.sonar.check.Rule;
import org.sonar.plugins.xquery.parser.XQueryParser;
import org.sonar.plugins.xquery.parser.XQueryTree;
import org.sonar.plugins.xquery.rules.CheckClasses;

/**
 * Checks for usage of text() XPath steps.
 * 
 * @since 1.0
 */
@Rule(
    key = XPathSubExpressionsInPredicateCheck.RULE_KEY,
    name = "Avoid XPath Sub-expressions in XPath Predicates",
    description = "Watch out for expressions like '[foo/bar]' or '[foo[bar]]' " +
            "because they can sometimes be bad for performance. " +
            "If the result is static it can be bound to a variable.",
    priority = Priority.INFO
)
public class XPathSubExpressionsInPredicateCheck extends AbstractPredicateCheck {
    // TODO: Maybe change this so it only catches nested predicates - unless
    // alternate ways to do sub-expressions is found

    public static final String RULE_KEY = "XPathSubExpressionsInPredicate";
    private static final RuleKey RULE = RuleKey.of(CheckClasses.REPOSITORY_KEY, RULE_KEY);

    @Override
    public void enterExpression(XQueryTree node) {
        super.enterExpression(node);

        if (inPredicate() && XQueryParser.PathExpr == node.getType() && node.getValue().contains("/")) {
            // Create a violation for any XPath step expressions
            createIssue(RULE, node.getLine());

        } else if (inPredicate() && XQueryParser.Predicate == node.getType()) {
            // Create a violation for nested predicates
            createIssue(RULE, node.getLine());

        } else if (XQueryParser.Predicate == node.getType()) {
            // If we are entering a predicate then update the state
            enterPredicate();
        }
    }
}

File: src/main/java/org/sonar/plugins/xquery/checks/XPathSubExpressionsInPredicateCheck.kt
/*
 * © 2014 by Intellectual Reserve, Inc. All rights reserved.
 */

package org.sonar.plugins.xquery.checks

import org.antlr.v4.runtime.ParserRuleContext
import org.antlr.v4.runtime.tree.TerminalNode
import org.sonar.api.rule.RuleKey
import org.sonar.check.Priority
import org.sonar.check.Rule
import org.sonar.plugins.xquery.checks.XPathSubExpressionsInPredicateCheck.Companion.RULE_KEY
import org.sonar.plugins.xquery.parser.XQueryParser.PredicateContext
import org.sonar.plugins.xquery.parser.XQueryParser.RelativePathExprContext
import org.sonar.plugins.xquery.parser.getLine
import org.sonar.plugins.xquery.rules.CheckClasses

/**
 * Checks for usage of text() XPath steps.
 *
 * @since 1.0
 */
@Rule(key = RULE_KEY)
class XPathSubExpressionsInPredicateCheck : AbstractPredicateCheck() {

    override fun enterExpression(node: ParserRuleContext) {
        super.enterExpression(node)

        if (inPredicate() && node is RelativePathExprContext) {
            // Create a violation for any XPath step expressions
            if (node.children?.any { it is TerminalNode } == true) {
                createIssue(RULE, node.getLine())
            }

        } else if (inPredicate() && node is PredicateContext) {
            // Create a violation for nested predicates
            createIssue(RULE, node.getLine())

        } else if (node is PredicateContext) {
            // If we are entering a predicate then update the state
            enterPredicate()
        }
    }

    companion object {
        // TODO: Maybe change this so it only catches nested predicates - unless
        // alternate ways to do sub-expressions is found

        const val RULE_KEY = "XPathSubExpressionsInPredicate"
        private val RULE = RuleKey.of(CheckClasses.REPOSITORY_KEY, RULE_KEY)
    }
}